src/Database/LSMTree/Internal/RunReader.hs

```diff
@@ -64,18 +65,54 @@ data RunReader m fhandle = RunReader {
 new ::
      HasFS IO h
   -> HasBlockIO IO h
+  -> Maybe SerialisedKey  -- ^ offset
   -> Run.Run IO (FS.Handle h)
   -> IO (RunReader IO (FS.Handle h))
-new fs hbio readerRun = do
+new fs hbio mOffset readerRun = do
     readerKOpsHandle <-
       FS.hOpen fs (runKOpsPath (Run.runRunFsPaths readerRun)) FS.ReadMode
-    -- double the file readahead window (only applies to this file descriptor)
+    -- Double the file readahead window (only applies to this file descriptor)
     FS.hAdviseAll hbio readerKOpsHandle FS.AdviceSequential
-    readerCurrentEntryNo <- newPrimVar 0
-    -- load first page from disk, if it exists.
-    firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs readerKOpsHandle
-    readerCurrentPage <- newIORef firstPage
+
+    (page, entryNo) <- case mOffset of
+      Nothing -> do
+        -- Load first page from disk, if it exists.
+        firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs readerKOpsHandle
+        return (firstPage, 0)
+
+      Just offset -> do
+        -- Use the index to find the page number for the key (if it exists).
+        let Index.PageSpan pageNo pageEnd = Index.search offset index
+        -- Load first page from disk, if it exists.
+        firstPage <- fromMaybe emptyRawPage <$> readDiskPageAt fs pageNo readerKOpsHandle
+
+        case rawPageFindKey firstPage offset of
+          Just n ->
+            -- Found an appropriate index within the index's page.
+            return (firstPage, n)
+
+          -- The index said that the key, if it were to exist, would live on pageNo,
+          -- but then rawPageFindKey tells us that in fact
+          -- there is no key greater than or equal to the given offset on this page.
+          -- So what can we conclude from the given information?
+          -- The information tells us that the key does not exist, but that if it were to exist,
+          -- it would be between the last key in this page and the first key in the next page.
+          -- Thus the reader should be initialised to return keys starting from the next page.
+          Nothing | pageNo == pageEnd ->
+            -- If already at the last page, there is no page of content to return.
+            -- Hence, no need to call readDiskPageAt.
+            return (emptyRawPage, 0)
```

Comment:
I don't understand this case. Do you mean something like `pageNo >= sizeInPages index`? 

`pageEnd` is not the last page of the entire run, just the last page of the (potential) entry we are looking for. So `pageNo == pageEnd` means that we don't have to deal with a larger-than-page entry here.